pkgs/development/libraries/flutter-engine/package.nix

```diff
@@ -0,0 +1,177 @@
+## RFC: should we use "hostPackages" to get the packages we need for "nativeBuildInputs"?
+{ lib, stdenv, stdenvNoCC, hostPlatform, callPackage, fetchFromGitHub, fetchurl, writeText,
+  ninja, gnumake, patchelf, python3, clang-tools, pkg-config, openssh, git, gclient-wrapped }:
+{ sha256, version, runtimeMode }:
+with lib;
+let
+  flutter-deps = filterAttrs (name: pkg: isAttrs pkg && hasAttr "outPath" pkg) (callPackage ./deps.nix {});
+
+  toolchainArch = if hostPlatform.isx86_64 then "amd64" else if hostPlatform.isAarch64 then "aarch64" else throws "Unsupported platform";
+  interpreter = "ld-linux-${hostPlatform.parsed.cpu.arch}.so.2";
+
+  src = stdenvNoCC.mkDerivation {
+    pname = "flutter-engine-src";
+    inherit version;
+
+    passthru = {
+      inherit flutter-deps;
+    };
+
+    nativeBuildInputs = [ patchelf python3 ];
+
+    src = fetchFromGitHub {
+      owner = "flutter";
+      repo = "engine";
+      rev = version;
+      inherit sha256;
+    };
+
+    buildroot = fetchFromGitHub {
+      owner = "flutter";
+      repo = "buildroot";
+      rev = "6af51ff4b86270cc61517bff3fff5c3bb11492e1";
+      sha256 = "sha256-03jr3H1RvzGfa0rBPZ1rtNpmieKzDjDgBsrZGaj7vuI=";
+    };
+
+    toolchain = fetchurl {
+      url = "https://commondatastorage.googleapis.com/chrome-linux-sysroot/toolchain/79a7783607a69b6f439add567eb6fcb48877085c/debian_sid_${toolchainArch}_sysroot.tar.xz";
+      sha256 = "sha256-+jooJ4YcWHUfD5AT+Yhe0nYQNBl7IvJl2dRc+d3Yl+g=";
+    };
+
+    dontUnpack = true;
+    dontConfigure = true;
+    dontBuild = true;
+
+    ## FIXME: use "flutter-deps" and the arguments which the OG script takes to generate the git revision
+    gitrev = writeText "git-revision.py" ''
+      #!${python3}/bin/python3
+
+      print("${fakeHash}")
+    '';
+
+    dartPackageConfig = writeText "dart-package-config.json" ''
+      {
+        "configVersion": 2,
+        "generator": "tools/generate_package_config.dart",
+        "packages": []
+      }
+    '';
+
+    binaryFixes = [
+      "src/third_party/dart/tools/sdks/dart-sdk/bin/dart"
+      "src/flutter/third_party/gn/gn"
+    ];
+
+    shellFixes = [
+      "src/third_party/dart/tools/sdks/dart-sdk/bin/dart2js"
+      "src/flutter/tools/gn"
+    ];
+
+    ## FIXME: instead of patching a bunch of binaries, maybe we copy from nixpkgs?
+    installPhase = ''
+      mkdir -p $out
+      cp -r -P --no-preserve=mode,ownership $buildroot $out/src
+      cp -r -P --no-preserve=mode,ownership $src $out/src/flutter
+
+      ${concatStringsSep "\n" (attrValues (mapAttrs (name: src: ''
+        if [[ -d ${src} ]]; then
+          mkdir -p $out/${name}
+          cp -r -P --no-preserve=mode,ownership ${src}/* $out/${name}
+          find ${src} -type f -name '.*' | xargs -I {} cp --no-preserve=mode,ownership {} $out/${name}
+        else
+          cp -P --no-preserve=mode,ownership ${src} $out/${name}
+        fi
+      '') flutter-deps))}
```

Comment:
why so complicated?
ideally just symlink the sources (read-only)
if we need write-access, just

```
mkdir -p $(dirname $out/${name})
cp -r ${src} $out/${name}
chmod -R +w $out/${name}
```

Reply:
> why so complicated?

because `cp -r ${src} $out/${name}` breaks when the folder `$out/${name}` exists...

could be solved with `rsync` but this is simpler

pkgs/development/libraries/flutter-engine/package.nix

```sh
    ## FIXME: instead of patching a bunch of binaries, maybe we copy from nixpkgs?  
    installPhase = ''
      mkdir $out
      # copy files and make them writable. "copy read write"
      function cprw {
        # TODO: do we really need "cp -P"?
        if [ -d "$1" ]; then
          mkdir -p "$2"   
          (
            shopt -s dotglob
            cp -r -P "$1/"* "$2"
          )
        else   
          mkdir -p "$(dirname "$2")"
          cp -P "$1" "$2"
        fi
        chmod -R +w "$2"
      }
      cprw $buildroot $out/src
      cprw $src $out/src/flutter

      ${concatStringsSep "\n" (attrValues (mapAttrs (name: src: ''
        cprw ${src} $out/${name}
      '') flutter-deps))}
```

Reply:
> because `cp -r ${src} $out/${name}` breaks when the folder `$out/${name}` exists...

This never will happen because of how `flutter-deps` is a set of attributes and because of the if directory function. The other sources also won't have the same directory name inside of each other.

Reply:
> The other sources also won't have the same directory name inside of each other.

there are 2 deps

```
src/third_party/boringssl
src/third_party/boringssl/src
```

Reply:
That's not what I meant, `src/third_party/boringssl`'s fetch will never have `src` already inside of it because we have to fetch `src/third_party/boringssl/src` ourselves.